Grouping Top N queries documentation (#8173)
opensearch-project/documentation-website: _observing-your-data/query-insights/grouping-top-n-queries.md

---
layout: default
title: Grouping top N queries
parent: Query insights
nav_order: 20
---

# Grouping top N queries
**Introduced 2.17**
{: .label .label-purple }

> Monitoring the [top N queries]({{site.url}}{{site.baseurl}}/observing-your-data/query-insights/top-n-queries/) helps identify the most resource-intensive queries based on latency, CPU, and memory usage over a specified time window. However, if a single computationally-expensive query is executed multiple times, it can occupy all top N query slots, potentially preventing other expensive queries from appearing in the list. To address this issue, you can group similar queries, gaining insight into various high-impact query groups.

Reviewer: ```suggestion
Monitoring the [top N queries]({{site.url}}{{site.baseurl}}/observing-your-data/query-insights/top-n-queries/) can help you to identify the most resource-intensive queries based on latency, CPU, and memory usage in a specified time window. However, if a single computationally expensive query is executed multiple times, it can occupy all top N query slots, potentially preventing other expensive queries from appearing in the list. To address this issue, you can group similar queries, gaining insight into various high-impact query groups.
```

> 

Reviewer: Line above: FYI, no hyphen in a compound adjective if the first word is an adverb ending in -ly

> Starting with OpenSearch version 2.17, top N queries can be grouped by `similarity`, with additional grouping options planned for future releases. 

Reviewer: ```suggestion
Starting with OpenSearch version 2.17, the top N queries can be grouped by `similarity`, with additional grouping options planned for future version releases. 
```

## Grouping queries by similarity

> Grouping queries by `similarity` organizes them based on the query structure, stripping out everything except the core query operations.

Reviewer: ```suggestion
Grouping queries by `similarity` organizes them based on the query structure, removing everything except the core query operations.
```

      ],
      "query_string": {
        "query": "search query"
      }
    }
  }
}
```

Has the following corresponding query structure:

```c
bool
  must
    exists
  query_string
```

When queries share the same query structure, they are grouped together, ensuring that all similar queries belong to the same group.

## Aggregate metrics per group

In addition to retrieving latency, CPU, and memory metrics for individual top N queries, you can obtain aggregate statistics for the
top N query groups. For each query group, the response includes the following statistics:
- The total latency, CPU usage, or memory usage (depending on the configured metric type) 
- The total query count

Using these statistics, you can calculate the average latency, CPU usage, or memory usage for each query group. 
The response also includes one example query from the query group. 

## Configuring query grouping

Before you enable query grouping, you must enable top N query monitoring for a metric type of your choice. For more information, see [Configuring top N query monitoring]({{site.url}}{{site.baseurl}}/observing-your-data/query-insights/top-n-queries/#configuring-top-n-query-monitoring).

> To configure grouping for Top N queries, follow these steps:

Reviewer: ```suggestion
To configure grouping for top N queries, use the following steps.
```

### Step 1: Enable top N query monitoring 

> Ensure that Top N is enabled for at least one of the metrics: latency, CPU, or memory. For more information, see [Configuring Top N query monitoring]({{site.url}}{{site.baseurl}}/observing-your-data/query-insights/top-n-queries/#configuring-top-n-query-monitoring).

Reviewer: Should "top" be lowercase here?

> Ensure that Top N is enabled for at least one of the metrics: latency, CPU, or memory. For more information, see [Configuring Top N query monitoring]({{site.url}}{{site.baseurl}}/observing-your-data/query-insights/top-n-queries/#configuring-top-n-query-monitoring).

Reviewer 2: ```suggestion
Ensure that top N query monitoring is enabled for at least one of the metrics: latency, CPU, or memory. For more information, see [Configuring Top N query monitoring]({{site.url}}{{site.baseurl}}/observing-your-data/query-insights/top-n-queries/#configuring-top-n-query-monitoring).
```

> For example, to enable top N query monitoring by latency with default settings, send the following request:

Reviewer: ```suggestion
For example, to enable top N query monitoring by latency with the default settings, send the following request:
```

> 

Reviewer: Line above: "by" => "of"?

```json
PUT _cluster/settings
{
  "persistent" : {
    "search.insights.top_queries.latency.enabled" : true
  }
}
```
{% include copy-curl.html %}

### Step 2: Configure query grouping

Set the desired grouping method by updating the following cluster setting:

```json
PUT _cluster/settings
{
  "persistent" : {
    "search.insights.top_queries.group_by" : "similarity"
  }
}
```
{% include copy-curl.html %}

> The default value for the `group_by` setting is `none`, which disables grouping. As of OpenSearch 2.17, the supported values for group_by are `similarity` and `none`.

Reviewer: ```suggestion
The default value for the `group_by` setting is `none`, which disables grouping. As of OpenSearch 2.17, the supported values for `group_by` are `similarity` and `none`.
```

> ### Step 3 (Optional): Set a limit on the number of query groups to monitor

Reviewer: ```suggestion
### Step 3 (Optional): Limit the number of monitored query groups
```

> Optionally, you can set a limit on the number of query groups to monitor. Queries that are already part of the top N query list (the most resource-intensive queries) will not count toward that limit. Essentially, the maximum applies only to other query groups, and the top N queries are tracked separately. This helps manage the tracking of query groups based on workload and query window size. 

Reviewer: ```suggestion
Optionally, you can limit the number of monitored query groups. Queries already included in the top N query list (the most resource-intensive queries) will not be considered in determining the limit. Essentially, the maximum applies only to other query groups, and the top N queries are tracked separately. This helps manage the tracking of query groups based on workload and query window size. 
```

          "taskId": 64,
          "parentTaskId": 63,
          "nodeId": "ZbINz1KFS1OPeFmN-n5rdg",
          "taskResourceUsage": {
            "cpu_time_in_nanos": 12161000,
            "memory_in_bytes": 473456
          }
        },
        {
          "action": "indices:data/read/search",
          "taskId": 63,
          "parentTaskId": -1,
          "nodeId": "ZbINz1KFS1OPeFmN-n5rdg",
          "taskResourceUsage": {
            "cpu_time_in_nanos": 293000,
            "memory_in_bytes": 3216
          }
        }
      ],
      "indices": [
        "my_index"
      ],
      "labels": {},
      "search_type": "query_then_fetch",
      "measurements": {
        "latency": {
          "number": 43,
          "count": 3,
          "aggregationType": "AVERAGE"
        }
      }
    }
  ]
}
```

</details>

## Response fields

The response includes the following fields.

Field | Data type        | Description
:--- |:---| :---
`top_queries` | Array | The list of top query groups.
`top_queries.timestamp` | Integer          | The execution timestamp for the first query in the query group.
`top_queries.source` | Object           | The first query in the query group.

> `top_queries.phase_latency_map` | Object           | The phase latency map for the first query in the query group. The map includes the times in milliseconds the query spent in `expand`, `query`, and `fetch` phases.

Reviewer: ```suggestion
`top_queries.phase_latency_map` | Object           | The phase latency map for the first query in the query group. The map includes the amount of time, in milliseconds, that the query spent in the `expand`, `query`, and `fetch` phases.
```

`top_queries.total_shards` | Integer          | The number of shards on which the first query was executed.
`top_queries.node_id` | String           | The node ID of the node that coordinated the execution of the first query in the query group.

> `top_queries.query_hashcode` | String           | The hashcode that uniquely identifies the query group. This is essentially the hash of the [query structure](#grouping-queries-by-similarity).

Reviewer 2: ```suggestion
`top_queries.query_hashcode` | String           | The hash code that uniquely identifies the query group. This is essentially the hash of the [query structure](#grouping-queries-by-similarity).
```

> `top_queries.task_resource_usages` | Array of objects | The resource usage breakdown for the various tasks for the first query in the query group.

Reviewer: ```suggestion
`top_queries.task_resource_usages` | Array of objects | The resource usage breakdown for the various tasks in the first query in the query group.
```

> `top_queries.indices` | Array | The indexes which the first query in the query group is searching.

Reviewer: ```suggestion
`top_queries.indices` | Array | The indexes searched by the first query in the query group.
```

`top_queries.labels` | Object           | Used to label the top query.

> `top_queries.search_type` | String           | The search request execution type (`query_then_fetch` or `dfs_query_then_fetch`). For more information, see the `search_type` parameter in the [Search API documentation]({{site.url}}{{site.baseurl}}/api-reference/search/#url-parameters).

Reviewer: Should "section" follow "parameter"? The reader can't literally see the parameter.

Reviewer 2: This is referring to a table that lists all parameters, and the `search_type` parameter is one of the rows in that table.

`top_queries.measurements` | Object           | The aggregate measurements for the query group.
`top_queries.measurements.latency` | Object           | The aggregate latency measurements for the query group.

> `top_queries.measurements.latency.number` | Integer          | The total latency for the query group.

Reviewer: Should "value" follow "latency"?

`top_queries.measurements.latency.count` | Integer          | The number of queries in the query group.

> `top_queries.measurements.latency.aggregationType` | String           | The aggregation type for the current entry. If grouping by similarity is enabled, then `aggregationType` is `AVERAGE`. If not enabled, then `aggregationType` is `NONE`. 

Reviewer: ```suggestion
`top_queries.measurements.latency.aggregationType` | String           | The aggregation type for the current entry. If grouping by similarity is enabled, then `aggregationType` is `AVERAGE`. If it is not enabled, then `aggregationType` is `NONE`. 
```
